ath79: Add support for COMFAST CF-E320N v2

rabin-io opened the pull request:
COMFAST CF-E320N v2 is a Ceiling AP with PoE support, based on
Qualcomm/Atheros QCA9531.

Short specification:

    2x 10/100 Mbps Ethernet, with 24v PoE support
    64 MB of RAM (DDR2)
    16 MB of FLASH (SPI)
    2T2R 2.4 GHz, 802.11b/g/n
    Built-in 2x 3 dBi antennas
    Output power (max): 110 mW (19 dBm)

Flash instruction:

Original firmware is based on OpenWrt.
Use sysupgrade image directly in vendor GUI.

Based on: https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=31952dbd1c40a2cd015ef435bd64fa9b2bb87eec

adschm commented:
So, what is WIP here?

rabin-io commented:
Need someone else to review the PR :)
but I see it's part of the pipeline already.

adschm commented:
So, now you removed [WIP], but told it's still WIP, and still we don't know why it actually was/is considered WIP ...

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e320n-v2.dts:
```diff
@@ -0,0 +1,165 @@
+};
+
+&wmac {
+	status = "okay";
```
Please add an empty line after status.

adschm replied:
no line added yet ...

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -134,6 +134,15 @@ comfast,cf-e314n-v2)
+	ucidef_set_led_rssi "rssilow" "RSSILOW" "red:rssilow" "wlan0" "1" "100"
+	ucidef_set_led_rssi "rssimediumlow" "RSSIMEDIUMLOW" "red:rssimediumlow" "wlan0" "26" "100"
+	ucidef_set_led_rssi "rssimediumhigh" "RSSIMEDIUMHIGH" "green:rssimediumhigh" "wlan0" "51" "100"
+	ucidef_set_led_rssi "rssihigh" "RSSIHIGH" "green:rssihigh" "wlan0" "76" "100"
```
This is wrong. Did you not test it?

rabin-io replied:
Yes, it is, it's a common template for all the comfast devices. 

And I'm not sure what are you asking me to change in the DTS file. If you can please expand your answer. 

Thanks

adschm replied:
The names you use here do not exist in DTS. So, it obvious you just copy/pasted it without knowing how it works and without having tested it.

rabin-io replied:
> MAC address info in the commit message is also still missing: https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses
> 
> And please squash your changes into a single commit and force-push them to your branch, so this PR is updated.

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: eth0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel master br-lan state DOWN mode DEFAULT group default qlen 1000
    link/ether 20:0d:b0:87:3f:71 brd ff:ff:ff:ff:ff:ff
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN mode DEFAULT group default qlen 1000
    link/ether 20:0d:b0:87:3f:6f brd ff:ff:ff:ff:ff:ff
6: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP mode DORMANT group default qlen 1000
    link/ether 20:0d:b0:87:3f:70 brd ff:ff:ff:ff:ff:ff
```

Can I assume it is like this ?

- eth0 = xx:71
- eth1 = xx:6f
- wlan0 = xx:70

I tried to find the info from the [original firmware](http://dl.comfast.cn/firmware/CF-E320NV2-V2.6.0.3.zip), but I didn't know how to extract the information from it.

adschm replied:
Typically, one would get the OEM assignment from a device with OEM firmware installed. Then, you could install OpenWrt on the device and look for the addresses in the usual locations (art 0x0, 0x6, 0x1002):
https://openwrt.org/docs/guide-developer/mac.address

This way you can relate the assignment from OEM to where you find these addresses, and finally compile the MAC address table I'm looking for. This is not at all possible without any OEM information.

One might also try to extract the same information from the OEM firmware, but it's not that easy.

Looking at your reported addresses here, you would have :6f for LAN and :71 for WAN, which would be the opposite compared to ar71xx. However, that doesn't help much, as we have three open questions related to that:
1. Is the assignment on ar71xx correct?
2. Is your assignment of LAN/WAN correct?
3. Is your setup of the &eth* DT nodes correct?

We do not have sufficient information yet to answer all these questions.

Unfortunately, the firmware download doesn't work for me either.

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -583,6 +583,16 @@ define Device/comfast_cf-e314n-v2
+  DEVICE_VENDOR := COMFAST
+  DEVICE_MODEL := CF-E320N
+  DEVICE_VARIANT := v2
+  DEVICE_PACKAGES := rssileds kmod-leds-gpio kmod-usb-core kmod-usb2
```
kmod-usb-core is selected automatically and can be dropped here.

And I don't think we need kmod-leds-gpio either.

adschm replied:
Hmm, and I don't see USB here at all ...

rabin-io replied:
I removed the USB and GPIO modules

rabin-io commented:
> So, now you removed [WIP], but told it's still WIP, and still we don't know why it actually was/is considered WIP ...

Since my only way to check it was by compiling and flashing, and hoping no magic smoke will came out.


I do not have the knowledge and experience to test it more thoroughly.

neheb commented on target/linux/ath79/dts/qca9531_comfast_cf-e320n-v2.dts:
```diff
@@ -0,0 +1,165 @@
+		pinctrl-0 = <&jtag_disable_pins &led_rssilow_pin &led_rssimediumhigh_pin &led_rssihigh_pin>;
+
+		wan {
+			label = "red:wan";
```
label property has been deprecated upstream. Please use color and function to specify names.

adschm replied:
No, we do not support color/function properly yet.

https://github.com/openwrt/openwrt/pull/3437

adschm commented:
MAC address info in the commit message is also still missing:
https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e320n-v2.dts:
```diff
@@ -0,0 +1,166 @@
+
+		wan {
+			label = "red:wan";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
```
In ar71xx, wan, lan and wlan LEDs are set up with GPIO_ACTIVE_HIGH. Who is right?

rabin-io replied:
Yes, you are right, it should be HIGH based on [old](https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=31952dbd1c40a2cd015ef435bd64fa9b2bb87eec
) commit.

![image](https://user-images.githubusercontent.com/5116363/139599279-c359d383-86bc-46da-a044-4c8240989d09.png)

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e320n-v2.dts:
```diff
@@ -0,0 +1,166 @@
+		};
+
+		rssilow {
+			label = "red:signal1";
```
Either you use red:rssilow here or rename the node to "signal1" ...

rabin-io replied:
OK.

Should I change all the others as well ?

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e320n-v2.dts:
```diff
@@ -0,0 +1,166 @@
+
+	led_rssihigh_pin: pinmux_rssihigh_pin {
+		pinctrl-single,bits = <0x10 0x0 0x000000ff>;
+	};
```
This sets 11, 14 and 16, but what about 12?

ar71xx also has:

> #define CF_E320N_V2_GPIO_HEADER_J9_2	12
> 	ath79_gpio_output_select(CF_E320N_V2_GPIO_HEADER_J9_2, 0);

adschm commented:
I miss wdt setup, which is done in ar71xx:

>		#define CF_E320N_V2_GPIO_EXT_WDT	13
>
>		gpio_request_one(gpio_wdt, GPIOF_OUT_INIT_HIGH,
>				 "PT7A7514 watchdog");
>
>		setup_timer(&gpio_wdt_timer, gpio_wdt_toggle, gpio_wdt);
> 		gpio_wdt_toggle(gpio_wdt);

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -134,6 +134,15 @@ comfast,cf-e314n-v2)
 	;;
+comfast,cf-e320n-v2)
+	ucidef_set_led_switch "lan" "LAN" "green:lan" "switch0" "0x04"
+	ucidef_set_led_switch "wan" "WAN" "red:wan" "switch0" "0x02"
```
Is this really working for both LEDs?

adschm commented on target/linux/ath79/dts/qca9531_comfast_cf-e320n-v2.dts:
```diff
@@ -0,0 +1,166 @@
+		rssimediumhigh {
+			label = "green:signal3";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
```
Okay, you inserted these now. They still don't fit what you have in 01_leds.

This tells me that you don't actually know what you are doing here.

ar71xx does not have them either.

Do these LEDs actually exist?

rabin-io committed:
ath79: Add support for COMFAST CF-E320N v2

COMFAST CF-E320N v2 is a Ceiling AP with PoE support, based on
Qualcomm/Atheros QCA9531.

Specifications
==============
- 650/393/216 MHz (CPU/DDR/AHB)
- 2x 10/100 Mbps Ethernet, both with PoE support
- 64 MB of RAM (DDR2)
- 16 MB of FLASH
- 2T2R 2.4 GHz, up to 22 dBm
- 2x internal antennas
- 1x RGB LED, 1x button
- UART (J1), GPIO (J9) and USB (J2) headers on PCB
- external watchdog (Pericon Technology PT7A7514)

Flash instruction:

Original firmware is based on OpenWrt.
Use sysupgrade image directly in vendor GUI.

Past work on this device: https://git.openwrt.org/?p=openwrt/openwrt.git;a=commitdiff;h=31952dbd1c40a2cd015ef435bd64fa9b2bb87eec

Signed-off-by: Rabin Yasharzadehe <rabin@rabin.io>

rabin-io commented:
Hi @adschm and @neheb 

The work on this branch is not based on any experience in embedded systems, and I do not know what I am doing exactly. I'm just trying to restore support for a previously supported device, and I do so based on templates of other devices with the values from the above COMMIT.

I hoped I can help with something, by doing so. But if I create more mess than help, feel free to close the PR and I will delete the PR.

adschm commented:
> Hi @adschm and @neheb
> 
> The work on this branch is not based on any experience in embedded systems, and I do not know what I am doing exactly. I'm just trying to restore support for a previously supported device, and I do so based on templates of other devices with the values from the above COMMIT.

That's okay, it will just take a little longer and require some help. But you should still try to think what you are doing.
If you add LED definitions (the block is called "leds"), you should think about whether these LEDs are there or not. And you should test whether they are actually working. Same goes for buttons.

It's not just filling in ACTIVE_HIGH somewhere now that I've told you so, but testing whether the device still boots properly and the button actually does what it's supposed to.

For actually finding out the correct GPIO values, you might also be interested in the lower part of https://openwrt.org/docs/techref/hardware/port.gpio

grobian commented:
I can confirm these patches (applied to 21.02.2) work, CF-320Nv2 is fully operational with these, thanks!

yumarca commented:
hello great job I was also looking for to port my comfast320n but I see you made it for 8mb flash but mine is 16mb. Is it possible to make the 16Mb version ?

neheb commented on target/linux/ath79/dts/qca9531_comfast_cf-e320n-v2.dts:
```diff
@@ -0,0 +1,166 @@
+&wmac {
+	status = "okay";
+
+	mtd-cal-data = <&art 0x1000>;
```
mtd-cal-data is gone and can no longer be used.

yumarca commented:
I increased the firmware is it ok ? Also made image size 15360k
define Device/comfast_cf-e320n-v2
  SOC := qca9531
  DEVICE_VENDOR := COMFAST
  DEVICE_MODEL := CF-E320N
  DEVICE_VARIANT := v2
  DEVICE_PACKAGES := rssileds
  IMAGE_SIZE := 15360k
endef
TARGET_DEVICES += comfast_cf-e320n-v2

&spi {
	status = "okay";

	flash@0 {
		compatible = "jedec,spi-nor";
		reg = <0>;
		spi-max-frequency = <25000000>;

		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "u-boot";
				reg = <0x000000 0x010000>; /* 64KB */
				read-only;
			};

			art: partition@10000 {
				label = "art";
				reg = <0x010000 0x010000>; /* 64KB */
				read-only;
			};

			partition@20000 {
				compatible = "denx,uimage";
				label = "firmware";
				reg = <0x020000 0xf50000>; /* Increased to 15MB */
			};

			partition@f70000 {
				label = "configs";
				reg = <0xf70000 0x010000>; /* 64KB */
				read-only;
			};

			partition@f80000 {
				label = "nvram";
				reg = <0xf80000 0x010000>; /* 64KB */
				read-only;
			};
		};
	};
};
